### ISSUE TITLE
Drop support for unmaintained Ruby versions

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor#2764]
Going forward, Asciidoctor will only support versions of CRuby which are still actively maintained (i.e., receiving updates) according to https://www.ruby-lang.org/en/downloads/. That also means Asciidoctor will only support versions of JRuby that are aligned with these versions of CRuby.

Support equates to which versions of Ruby we test against.

At the time of filing this issue, this includes CRuby 2.2 and up, which means JRuby 9.0.x or JRuby 9.1.x.

What we won't do is set a minimum version of Ruby in the gemspec. The reason for this decision is that it's not a very friendly way to enforce a version, in my experience. It can block installation of gems which depend on asciidoctor that are designed to run on older versions of Ruby, which is an irritating situation and one which cannot be worked around. It's better to clearly state that using an unsupported version of Ruby is an "at your own risk" sort of thing. Enforcement through policy is the route I prefer to take. There's no guarantee Asciidoctor will actually work in those scenarios, but there's no guarantee it won't either.

**Comment by mojavelinux** [comment]
Since we'll be dropping Ruby < 2 (I know...those versions are ancient, but we've been very nice), here are things that can now be changed or refactored:

* remove core_ext patches (for Ruby < 2)
* switch to modern Hash syntax (e.g., `key: value`)
* remove places where magic regexp vars are backed up (look for `m = $~`)
* remove unicode escape sequence hacks (e.g., drop `150.chr`)
* remove `require 'rubygems'` from bin script
* use require_relative instead of modifying the $LOAD_PATH
* use inherit flag of const_get to look for constant in current module only
* use truncate for number precision calculations
* don't enclose method call in brackets when assigning multiple variables
* mark private methods as private (and perhaps use protected as well)
* replace replace .lines.entries with .lines in tests
* consider reading files using r:UTF-8 mode
* drop legacy regexp constants
* remove force unicode line length hack
* remove qualified name lookup patch in extensions
* use gsub w/ hash replacement

**Comment by mojavelinux** [comment]
We will have to be conscious that Opal still relies on some of the patches for old versions of Ruby. We'll just need to tease those apart.

**Comment by mojavelinux** [comment]
The new minimum versions are C Ruby 2.3 and up and JRuby 9.1 and up. The policy is to follow the lifecycle of the runtimes. (After all, why should we support what they don't support?)

**Comment by mojavelinux** [comment]
We should also look into enabling [frozen string literal](https://bugs.ruby-lang.org/issues/11473), which will become the default in Ruby 3.

**Comment by mojavelinux** [comment]
@Mogztter It looks like Opal doesn't support File.absolute_path. It's the same as File.expand_path, except it doesn't expand `~`. Since Opal doesn't do that anyway, it might be as simple as adding an alias.

**Comment by Mogztter** [comment]
@mojavelinux Indeed the Node.js implementation does not expand `~` (tilde) and it's unclear what the expected behavior would be in a browser environment. In Node.js we could replace non-escaped leading `~` with `require('os').homedir()` ?

> Since Opal doesn't do that anyway, it might be as simple as adding an alias.

:+1: 
https://github.com/opal/opal/issues/1922

**Comment by mojavelinux** [comment]
I realized that most of the use of expand_path I'm replacing would only affect compilation, just FYI.

**Comment by mojavelinux** [comment]
Ruby has introduced many new facilities for modifying the encoding of strings. One of those is the ability to set the string encoding when reading the contents of a file (e.g., `mode: 'rb:utf-8'`). But it goes even further. Ruby can also honor the encoding specified by the BOM (byte order mark). That functionality is activated by setting the mode to `rb:bom|utf-8`. That says "read the file (without messing with newlines), use the encoding specified by the BOM, if present, otherwise use UTF-8".

At first, I thought that feature would allow us to get rid of the BOM processing in Asciidoctor. As tempting as it may seem, the problem is that Asciidoctor accepts input from a variety of sources, so we aren't always in control of reading the input from a file (and we cann't get the read mode from the open file descriptor). And there's no way for us to know if the string was properly encoded and the BOM processed. There is a very narrow use case we could accommodate, but the amount of extra code we'd have to put in place would likely offset the benefit (though we still might consider it).

There's also the problem that `rb:bom|utf-8` doesn't work with URIs. So we'd still have to have our own encoding and BOM processing logic in those cases.

Where we can benefit is that we can assume that the input string is UTF-8 encoded (and always read it as such). Although we'd still check for the BOM, we know that the string is originally in UTF-8 when reencoding, which simplifies the logic.

Long story short, we can the `rb:utf-8` mode when reading files, except for Opal, which will use `r:utf-8` (since JavaScript always uses a breed of UTF when encoding strings).

Another change we will make is that we will read files whole instead of reading it into an array of lines. We want to do this since we have to walk all the lines to prepare them for parsing anyway (to encode as UTF-8 and remove trailing whitespace). There's no reason to walk the lines twice.

**Comment by mojavelinux** [comment]
It's also time to move away from modifying the $LOAD_PATH for requiring paths in the gem. There are two possible solutions:

1. Use require_relative
2. Use require, but modify the $LOAD_PATH in development and testing

Both solutions work equally well. The downside of (2) is that it doesn't address the issue of the leaky require. If you add a file somewhere on the $LOAD_PATH that matches a path in the Asciidoctor gem, Ruby might load that file instead of the one from the gem (depending on your $LOAD_PATH order). I consider this a bad thing. There are ways in Ruby to override / monkeypatch and this isn't the right way to go about it.

I think require_relative actually makes the behavior the most consistent between the testing and installed environments. Once you require the entry path, everything else gets required relative to that point, regardless of whether or not the gem is installed.

If we do use require_relative, we will need to still use require for the Asciidoctor.js parts. Otherwise, it will look for the files in the gem instead of on the modified $LOAD_PATH in the Asciidoctor.js build.

**Comment by mojavelinux** [comment]
> we can use the rb:utf-8 mode when reading files

Update: I've since learned that it's also possible to specify the internal encoding using: `rb:utf-8:utf-8`. This ensures that when the content is read from the file, the resulting string is also UTF-8 encoded. Otherwise, we have problems.

**Comment by mojavelinux** [comment]
@Mogztter I found this note in the code:

> NOTE Opal can't call key? on ENV

It seems to be specific to the nodejs module in Opal.

**Comment by mojavelinux** [comment]
> set using Ruby's `-E :utf-8` option

Here's how that's used:

```
ruby -E utf-8:utf-8 $(which asciidoctor) doc.adoc
```

**Comment by skalee** [comment]
See: https://github.com/asciidoctor/asciidoctor/commit/905f406e452becdbab05339e510535406bc65e13#diff-55d040d8106797c32ac6bcee48adf84eR20

Is it really a good idea that `required_ruby_version` is considered only informational? This gem is actually won't work correctly on Ruby < 2.3 (e.g. https://travis-ci.org/lsegal/yard/jobs/512898313), neither is tested on older Rubies.

**Comment by mojavelinux** [comment]
I feel strongly that the gemspec should not declare the required Ruby version. That's because I disagree with how the gem resolution works. In particular, gem resolution does not allow the user to override the suggested constraints, leaving the user no option for using backports to provide compatibility. (I've been personally bit by this many times). This is especially problematic for Linux distributions, which may use their own patches to provide support for older versions of Ruby. So I'm sticking with my policy of documenting the constraints rather than hard coding them into the gemspec.

**Comment by mojavelinux** [comment]
FYI, I decided to publish a patch version of Asciidoctor that restores compatibility with older versions of Ruby 2. Although we don't test on those versions of Ruby, it wasn't difficult to at least be nice and allow the gem to run. I think this is a fair trade-off.
